Source/DafnyCore/Compilers/Compiler-java.cs

```diff
@@ -2344,61 +2344,78 @@ private class GenericArrayElementLvalue : ILvalue {
-            : new List<string> { "--create", "--main-class", simpleName, "--file", jarPath };
-        psi = PrepareProcessStartInfo("jar", args.Concat(classfiles));
-        psi.WorkingDirectory = Path.GetFullPath(Path.GetDirectoryName(targetFilename));
-        compileIsSuccess &= 0 == RunProcess(psi, outputWriter, "Error while creating jar file.");
-
-        // Keep the build artifacts if --spill-target-code is anything but 0
-        // But keep them for now if the option is 0, (because users and tests likely depend on it), but deprecate
-        if (false && DafnyOptions.O.SpillTargetCode == 0) {
-          System.IO.Directory.Delete(targetDirectory, true);
-        }
-        if (compileIsSuccess && DafnyOptions.O.CompileVerbose) {
-          outputWriter.WriteLine("Wrote " + (callToMain != null ? "executable" : "library") + " jar " + (simpleName + ".jar"));
+      if (!DafnyOptions.O.UseRuntimeLib) {
+        // If the built-in runtime library is used, unpack it so it can be repacked into the final jar
+        var libUnpackProcess = PrepareProcessStartInfo("jar", new List<String> { "xf", "DafnyRuntime.jar" });
+        libUnpackProcess.WorkingDirectory = Path.GetFullPath(Path.GetDirectoryName(targetFilename));
+        if (0 != RunProcess(libUnpackProcess, outputWriter, "Error while creating jar file (unzipping runtime library).")) {
+          return false;
         }
       }
-      return compileIsSuccess;
+
+      // Find all .class files
+      var classfiles = new List<String>();
+      foreach (string file in Directory.EnumerateFiles(targetDirectory, "*.class", SearchOption.AllDirectories)) {
+        classfiles.Add(Path.GetRelativePath(targetDirectory, file));
+      }
+
+      string simpleProgramName = Path.GetFileNameWithoutExtension(targetFilename);
+
+      // Create the jar, in the specified output directory (dafnyProgramName is an absolute path to the desired output file)
+      System.IO.Directory.CreateDirectory(Path.GetDirectoryName(dafnyProgramName));
+      string jarPath = Path.GetFullPath(Path.ChangeExtension(dafnyProgramName, ".jar"));
+      var args = callToMain == null ?
+          new List<string> { "cf", jarPath }
+          : new List<string> { "cfe", jarPath, simpleProgramName };
+      var jarCreationProcess = PrepareProcessStartInfo("jar", args.Concat(classfiles));
+      jarCreationProcess.WorkingDirectory = Path.GetFullPath(Path.GetDirectoryName(targetFilename));
+      if (0 != RunProcess(jarCreationProcess, outputWriter, "Error while creating jar file: " + jarPath)) {
+        return false;
+      }
+
+      // Keep the build artifacts if --spill-translation is true
+      // But keep them for legacy CLI if the -spillTargetCode is 0, (because users and tests likely depend on it), but deprecate
+      if (DafnyOptions.O.UsingNewCli && DafnyOptions.O.SpillTargetCode == 0) {
+        System.IO.Directory.Delete(targetDirectory, true);
+      }
+
+      // When verbose, report the generated artifact
```

Comment:
Comment seems obsolete

Reply:
Not obsolete but maybe too obvious to comment. Removed.